src/sys/mod.rs

```diff
@@ -180,6 +161,21 @@ impl Token {
 /// source and delegate the implementations to it.
 pub struct Poll {
     poller: Poller,
+
+    // It is essential for safe use of this type that the pointers passed in to
+    // the underlying poller API are properly managed. Each time an event source
+    // is registered, the token it passes in is Boxed and converted to a raw
+    // pointer to be passed to the polling system by FFI. This pointer is what's
+    // stored in the map. When the event source is re- or unregistered, the same
+    // raw pointer can then be converted back into the Box and dropped, safely
+    // deallocating it. To put it another way, we effectively "own" the Token
+    // memory on behalf of the underlying polling mechanism.
+    //
+    // All the platforms we currently support follow the rule that file
+    // descriptors must be "small", positive integers. This means we can use a
+    // VecMap which has that exact constraint for its keys. If that ever
+    // changes, this will need to be changed to a different structure.
+    tokens: VecMap<*mut Token>,
```

Comment:
Why `VecMap<*mut Token>` rather than `VecMap<Box<Token>>`, given the `VecMap` is effectively what owns the tokens?

Reply:
I thought about this, the issue is that
- turning the `Box` into a pointer consumes the `Box`, so you'd need to either
  - use an `Rc` (which I originally did but it was very verbose)
  - immediately convert the pointer back into a `Box`, store that, give a copy of the pointer to `epoll()`, and then what if you inadvertently drop it in some other way? It just becomes another safety liability.
- the pointer can't deallocate the data by itself, the worst thing it can do is be forgotten about or dropped, which means a leak rather than dangling pointers elsewhere. So you're holding onto something relatively safe, that can only do unsafe things inside an actual `unsafe` block

I think it might be better semantics to think of `epoll()` (or `kqueue()` etc) as owning the heap allocation here, rather than the `Poll` or the `VecMap` member. At least, in between `[re|un]register()` calls. Then it makes sense to keep only the pointer, and to do the `Box<->raw` conversion right at the boundary of the FFI call (almost, anyway).